src/services/telemetryService.ts

```diff
@@ -0,0 +1,131 @@
+import * as vscode from 'vscode';
+import Analytics from 'analytics-node';
+import { TelemetryEvent } from '../interfaces/telemetryEvent';
+import { Reporter } from '../reporter';
+import { Logger } from '../utils/logger';
+import { SegmentInitializer } from '../utils/segmentInitializer';
+import { TelemetryEventQueue } from '../utils/telemetryEventQueue';
+
+/* 
+  OPT_IN_STATUS == "true" if user Agreed
+  false when denied for telemetry data collection     
+*/
+const OPT_IN_STATUS = 'redhat.telemetry.optInRequested';
+
+export namespace TelemetryService {
+  let analyticsObject: Analytics | undefined;
+
+  /* 
+    provides subscription to custom segment key.
+    fallback to default segment key if no key provided via API
+  */
+  export function subscribeTelemetryService(
+    clientExtensionName: string
+  ): boolean {
+    Reporter.setClientExtensionName(clientExtensionName);
+    const CLIENT_SEGMENT_KEY: string | undefined = getClientSegmentKey(
+      clientExtensionName
+    );
+    analyticsObject = SegmentInitializer.initialize(CLIENT_SEGMENT_KEY);
+    return analyticsObject !== undefined;
+  }
+
+  function getClientSegmentKey(
+    clientExtensionName: string
+  ): string | undefined {
+    try {
+      const clientPackageJson = vscode.extensions.getExtension(
+        clientExtensionName
+      )?.packageJSON;
+      const clientSegmentKey = clientPackageJson['segmentWriteKey'];
+      Logger.log(`client Segment-Key : ${clientSegmentKey}`);
+      return clientSegmentKey;
+    } catch (error) {
+      Logger.log('Unable to get Client Segment-Key');
+    }
+    return undefined;
+  }
+
+  /* 
+    Collects telemetry data and pushes to a queue when not opted in
+    and to segment when user has opted for telemetry 
+  */
+  export function sendEvent(event: TelemetryEvent) {
+    Logger.log('Event received:');
+    Logger.log(event.uuid);
```

Comment:
These should be merged into single log call.

Reply:
done